projects/bootloader/scripts/can_datagram.py

```diff
@@ -0,0 +1,294 @@
+                                 *data_crc32])
+        # Update the crc32
+        crc32 = zlib.crc32(crc32_array)
+        crc32 = self._convert_to_bytearray(self, crc32, 4)
+
+        # Update the bytearray
+        return crc32
+
+
+class DatagramSender:
+    """A class that acts as a distributor for the Datagram class on a bus."""
+
+    def __init__(self, bustype="socketcan", channel=DEFAULT_CHANNEL,
+                 bitrate=CAN_BITRATE, receive_own_messages=False):
+        self.bus = can.interface.Bus(
+            bustype=bustype,
+            channel=channel,
+            bitrate=bitrate,
+            receive_own_messages=receive_own_messages)
+
+    def send(self, message):
+        """Sends the Datagrams."""
+
+        assert isinstance(message, Datagram)
+
+        chunk_messages = self._chunkify(message.serialize(), 8)
+        message_arbitration_id = CAN_ARBITRATION_ID
+        message_extended_arbitration = False
+
+        can_messages = [
+            can.Message(
+                arbitration_id=CAN_START_ARBITRATION_ID,
+                data=bytearray(),
+                is_extended_id=message_extended_arbitration)]
+
+        # Populate an array with the can message from the library
+        for chunk_message in chunk_messages:
+            can_messages.append(can.Message(arbitration_id=message_arbitration_id,
+                                            data=chunk_message,
+                                            is_extended_id=message_extended_arbitration))
+
+            # After the first message, set the arbitration ID to 0
+            message_arbitration_id = CAN_ARBITRATION_ID
+
+        # Send the messages
+        for msg in can_messages:
+            self.bus.send(msg)
+        print("{} messages were sent on {}".format(len(can_messages), self.bus.channel_info))
+
+    def _chunkify(self, data, size):
+        """This chunks up the datagram bytearray for easy iteration."""
+        return (data[pos:pos + size] for pos in range(0, len(data), size))
+
+
+class DatagramListener(can.BufferedReader):
+    """A class that handles a callback when a datagram is received."""
+
+    def __init__(self, callback):
+        """Registers the callback."""
+        assert callable(callback)
+        self.callback = callback
+        # Messages are stored in a dictionary where key = board ID, value = message
+        self.datagram_messages = {}
+
+        super().__init__()
+
+    def on_message_received(self, msg: can.Message):
+        """Handles message sent from boards on the CAN."""
+        super().on_message_received(msg)
+
+        board_id = (msg.arbitration_id & 0b11111100000) >> 5
+        start_message = (msg.arbitration_id & 0x10) >> 4
+
+        if start_message == 1:
+            # Reset the datagram message when receiving a start message
+            self.datagram_messages[board_id] = msg.data
+
+        if start_message == 0:
+            self.datagram_messages[board_id] += msg.data
```

Comment:
One robustness point: this will error out if a non-start message is sent before any start messages are sent. This is possible if e.g. we start listening in the middle of a datagram. So you want to check that `board_id in self.datagram_messages` before doing this (or attempting to receive the datagram below) in order to just ignore "partial" datagrams.